agent/eventhandler/task_handler.go

```diff
@@ -15,244 +15,323 @@ package eventhandler
 	event.Containers = append(event.Containers, handler.tasksToContainerStates[event.TaskARN]...)
 	delete(handler.tasksToContainerStates, event.TaskARN)
-}
 
-// addEvent prepares a given event to be sent by adding it to the handler's appropriate
-// eventList and remove the entry in tasksToEvents map
-func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
-	handler.taskHandlerLock.Lock()
-	defer handler.taskHandlerLock.Unlock()
-	seelog.Infof("TaskHandler, Adding event: %s", change.String())
-
-	taskEvents := handler.getTaskEventList(change)
-
-	taskEvents.eventListLock.Lock()
-	defer taskEvents.eventListLock.Unlock()
-
-	// Update taskEvent
-	taskEvents.events.PushBack(change)
-
-	if !taskEvents.sending {
-		taskEvents.sending = true
-		go handler.SubmitTaskEvents(taskEvents, client)
-	}
+	// Prepare a given event to be sent by adding it to the handler's appropriate
+	// eventList and remove the entry in tasksToEvents map
+	change := newSendableTaskEvent(*event)
+	taskEvents := handler.getTaskEventsUnsafe(change)
 
-	delete(handler.tasksToEvents, change.taskArn())
+	taskEvents.sendChange(change, client, handler)
 }
 
-// getTaskEventList gets the eventList from taskToEvent map
-func (handler *TaskHandler) getTaskEventList(change *sendableEvent) (taskEvents *eventList) {
-	taskEvents, ok := handler.tasksToEvents[change.taskArn()]
+// getTaskEventsUnsafe gets the eventList from taskToEvent map
+func (handler *TaskHandler) getTaskEventsUnsafe(change *sendableEvent) *taskSendableEvents {
+	taskARN := change.taskArn()
+	taskEvents, ok := handler.tasksToEvents[taskARN]
 	if !ok {
-		seelog.Debug("TaskHandler, collecting events for new task ", change)
-		taskEvents = &eventList{events: list.New(), sending: false}
-		handler.tasksToEvents[change.taskArn()] = taskEvents
+		taskEvents = &taskSendableEvents{
+			events:    list.New(),
+			sending:   false,
+			createdAt: time.Now(),
+			taskARN:   taskARN,
+		}
+		handler.tasksToEvents[taskARN] = taskEvents
+		seelog.Debugf("TaskHandler: collecting events for new task; event: %s; events: %s ",
+			change.toString(), taskEvents.toStringUnsafe())
 	}
 
 	return taskEvents
 }
 
 // Continuously retries sending an event until it succeeds, sleeping between each
 // attempt
-func (handler *TaskHandler) SubmitTaskEvents(taskEvents *eventList, client api.ECSClient) {
-	backoff := utils.NewSimpleBackoff(1*time.Second, 30*time.Second, 0.20, 1.3)
+func (handler *TaskHandler) submitTaskEvents(taskEvents *taskSendableEvents, client api.ECSClient, taskARN string) {
+	defer handler.removeTaskEvents(taskARN)
+
+	backoff := utils.NewSimpleBackoff(submitStateBackoffMin, submitStateBackoffMax,
+		submitStateBackoffJitterMultiple, submitStateBackoffMultiple)
 
 	// Mirror events.sending, but without the need to lock since this is local
 	// to our goroutine
 	done := false
-
+	// TODO: wire in the context here. Else, we have go routine leaks in tests
```

Comment:
Is this TODO expected to be fixed before merging?  goroutine leaks in tests can cause flaky tests.

Reply:
Not with this PR :(